### PR TITLE
Support multi trigger apps

### PR DESCRIPTION
**Pull request by rajatjindal**
This PR adds support for running Spin apps with multiple triggers. I opened the draft PR to be able to collect some initial feedback.

I've also added a testcase to demonstrate running the multi-trigger-app workload and will be working on adding the integration test for the same.

### CODE REVIEW

File: images/spin-multi-trigger-app/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM --platform=${BUILDPLATFORM} rust:1.73 AS build
```
Comment (radu-matei) [inline images/spin-multi-trigger-app/Dockerfile:1]: I feel strongly that we should test using `spin registry push`

Comment (rajatjindal) [reply images/spin-multi-trigger-app/Dockerfile:1]: yes, that is on my todo list for next week (also discuss and cover more k8s distros probably).

File: containerd-shim-spin/src/engine.rs [hunk 2]
```diff
@@ -182,57 +183,74 @@ impl SpinEngine {
...
-        self.run_trigger(&trigger_cmd, locked_app, app_source).await
+        self.run_trigger(
+            trigger_cmds.iter().map(|s| s.as_ref()).collect(),
+            locked_app,
+            app_source,
+        )
+        .await
     }
 
     async fn run_trigger(
```
Comment (Mossaka) [inline containerd-shim-spin/src/engine.rs:197]: I wonder, the fact that we are blocking the futures in `run_wasi()` call, how would the concurrent performance be like for running multiple triggers at the same time.

Comment (rajatjindal) [reply containerd-shim-spin/src/engine.rs:197]: Hi @Mossaka ,

I ran some tests for this (using k3d cluster on my Macbook M2 pro):

(ran using `hey -z 10s http://localhost:8083/hello`)

| Scenario  | http-throughput  |
| ------------- | ------------- |
| v0.13.1 + only http-trigger app  | ~ 89k/s  |
| multi-trigger-shim + only http-trigger receiving traffic  | ~89k/s |
| multi-trigger-shim + both redis and http-trigger receiving traffic  | ~82k/s |

File: images/spin-multi-trigger-app/Dockerfile (hunk 1 above)
Comment (rajatjindal) [reply images/spin-multi-trigger-app/Dockerfile:1]: there is a ticket for it already. I will pick it up next. https://github.com/spinkube/containerd-shim-spin/issues/37

File: containerd-shim-spin/src/engine.rs [hunk 3]
```diff
@@ -182,57 +183,74 @@ impl SpinEngine {
...
+            };
+
+            futures_list.push(f)
+        }
+
         info!(" >>> notifying main thread we are about to start");
-        f.await
+
+        // exit as soon as any of the trigger completes/exits
+        let (result, _, rest) = future::select_all(futures_list).await;
```
Comment (jsturtevant) [inline containerd-shim-spin/src/engine.rs:249]: Why as soon as one completes?  Would that mean if the result of one of the others failed we woudln't know about it?  should we be waiting for all triggers finish, or some type of queue to handle the various results?

Comment (rajatjindal) [reply containerd-shim-spin/src/engine.rs:249]: Hello James, 

this is to keep it in sync with the [Spin behavior](https://github.com/fermyon/spin/blob/86a58771091c02eaf40b5395a2a6bc711740f038/src/commands/up.rs#L222)

File: containerd-shim-spin/src/engine.rs (hunk 2 above)
Comment (Mossaka) [reply containerd-shim-spin/src/engine.rs:197]: okay that looks good. what's the throughput of the redis trigger in your last scenario?

File: containerd-shim-spin/src/engine.rs (hunk 3 above)
Comment (Mossaka) [reply containerd-shim-spin/src/engine.rs:249]: could we add a log to print which trigger failed / terminates?

**Review by Mossaka — COMMENTED** [review]
Mostly LGTM, there is one pending comment from James that I am interested in seeing resolved as well.

File: containerd-shim-spin/src/engine.rs (hunk 3 above)
Comment (rajatjindal) [reply containerd-shim-spin/src/engine.rs:249]: 👍 

```
time="2024-04-04T02:58:27.951588261Z" level=info msg=" >>> trigger type 'redis' exited"
```

Comment (rajatjindal) [reply containerd-shim-spin/src/engine.rs:249]: @jsturtevant - does that answer your question? (re: exit behavior). thanks.

File: containerd-shim-spin/src/engine.rs (hunk 2 above)
Comment (rajatjindal) [reply containerd-shim-spin/src/engine.rs:197]: with both http & redis receiving traffic concurrently, I am able to inject about 1050 msgs (and I assume that also equates to number of events being processed by spin app)

with just redis receiving the traffic, the throughput is about 1450 msgs.

File: containerd-shim-spin/src/engine.rs (hunk 3 above)
Comment (jsturtevant) [reply containerd-shim-spin/src/engine.rs]: > this is to keep it in sync with the [Spin behavior](https://github.com/fermyon/spin/blob/86a58771091c02eaf40b5395a2a6bc711740f038/src/commands/up.rs#L222)

Interesting!  I guess we should do the same as spin cli, thanks!

**Comment by rajatjindal** [comment]
Hi @kate-goldenring / @Mossaka , I have rebased this PR and it is ready for review again. thank you

**Review by Mossaka — APPROVED** [review]
LGTM

**Comment by Mossaka** [comment]
Need a rebase

**Comment by kate-goldenring** [comment]
@rajatjindal it looks like the multi trigger app failed to run in the CI (even though the test passed): https://github.com/spinkube/containerd-shim-spin/actions/runs/8619544944/job/23625457710?pr=53#step:8:1234

**Comment by rajatjindal** [comment]
Hi @kate-goldenring @Mossaka , this PR has been rebased now, and seems to be ready for final set of review/merge. 

Thank you so much for your patience while I got this PR into a stable state.

**Review by kate-goldenring — APPROVED** [review]

**Review by devigned — APPROVED** [review]
lgtm

**Pull request merged**